### PR TITLE
Qt gui widget

### PR DESCRIPTION
**Pull request by harrisonyei**
Add qt widget of main window and displace video

### CODE REVIEW

File: src/MyGLWidget.h [hunk 1]
```diff
@@ -0,0 +1,25 @@
+#pragma once
```
Comment (john0312) [inline src/MyGLWidget.h]: Please add license header. Please use define guard.

File: src/MyGLWidget.h [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+#include <QOpenGLWidget>
+#include <Qtimer>
+#include <opencv2\core.hpp>
+#include <queue>
+class MyGLWidget : public QOpenGLWidget
```
Comment (john0312) [inline src/MyGLWidget.h]: Please add a brief comment for the class.

File: src/MyGLWidget.h [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+	Q_OBJECT
+
+public:
+	MyGLWidget(QWidget *parent);
+	~MyGLWidget();
+
+	void setTexture(cv::Mat mat);
+	 
+protected:
+	void initializeGL();
```
Comment (john0312) [inline src/MyGLWidget.h]: Please add brief comment for each member function.

File: src/MyGLWidget.h [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+	void paintGL();
+	void resizeGL(int width, int height);
+	
+private:
+	cv::Mat lastFrame;
```
Comment (john0312) [inline src/MyGLWidget.h]: Please add brief comment for each member variable.

File: src/TwitchWindow.cpp [hunk 5]
```diff
@@ -0,0 +1,61 @@
+#include "TwitchWindow.h"
```
Comment (john0312) [inline src/TwitchWindow.cpp]: Please separate into a different PR.
Also, please avoid using the name "Twitch", it's a trademark.

File: src/MyGLWidget.h (hunk 4 above)
Comment (harrisonyei) [reply src/MyGLWidget.h]: Comment added.

File: src/DisplayGLWidget.cpp [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+
+#include "DisplayGLWidget.h"
+#include <iostream>
+#include <opencv2\imgproc.hpp>
+
+DisplayGLWidget::DisplayGLWidget(QWidget *parent)
+	: QOpenGLWidget(parent)
+{
+	// add timer to automatically update window.
+	// connect(&timer, SIGNAL(timeout()), this, SLOT(update()));
```
Comment (john0312) [inline src/DisplayGLWidget.cpp]: Please remove unused code.

File: src/DisplayGLWidget.cpp [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+	// timer.start(10);
+}
+
+DisplayGLWidget::~DisplayGLWidget()
```
Comment (john0312) [inline src/DisplayGLWidget.cpp]: Consider using
= default;

File: src/DiplayGLWidget.h [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+	// @param QWidet* parent
+	// Constructs a GLWidget under the parent qtwidget.
+	DisplayGLWidget(QWidget *parent);
+
+	// Destructor
+	~DisplayGLWidget();
+
+	// Sets current texture from given mat data.
+	// @param cv::Mat mat
+	void setTexture(cv::Mat mat);
```
Comment (john0312) [inline src/DiplayGLWidget.h]: Please use const ref or pointer.

File: src/DisplayGLWidget.cpp [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+	glOrtho(-w, w, -h, h, -10, 10);
+
+	// set modelview transformation
+	glMatrixMode(GL_MODELVIEW);
+	glLoadIdentity();
+
+	glEnable(GL_TEXTURE_2D);
+
+	// bind shader
+	//this->commom_shader->Use();
```
Comment (john0312) [inline src/DisplayGLWidget.cpp]: Please remove unused code. You can add a TODO.

File: src/DisplayGLWidget.cpp [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+
+	// bind texture
+	glBindTexture(GL_TEXTURE_2D, textureID);
+
+	// if frame data is not empty
+	if (!lastFrame.empty()) {
+		//cv::Mat frame(100, 100, CV_8UC3, cv::Scalar(0, 0, 255));
+		int width = lastFrame.cols;
```
Comment (john0312) [inline src/DisplayGLWidget.cpp]: Consider using:
const int&

File: src/DisplayGLWidget.cpp [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+
+	// bind shader
+	//this->commom_shader->Use();
+
+	// bind texture
+	glBindTexture(GL_TEXTURE_2D, textureID);
+
+	// if frame data is not empty
+	if (!lastFrame.empty()) {
+		//cv::Mat frame(100, 100, CV_8UC3, cv::Scalar(0, 0, 255));
```
Comment (john0312) [inline src/DisplayGLWidget.cpp]: Please remove unused code.

File: src/DisplayGLWidget.cpp [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+		int width = lastFrame.cols;
+		int height = lastFrame.rows;
+
+		unsigned char* image = lastFrame.data;
```
Comment (john0312) [inline src/DisplayGLWidget.cpp]: Please add const if possible.

File: src/DisplayGLWidget.cpp [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+		// if data is not nullptr
+		if (image)
+		{
+			// wrtie texture data from given image data.
+			glTexImage2D(GL_TEXTURE_2D, 0, GL_RGB, width, height, 0, GL_RGB, GL_UNSIGNED_BYTE, image);
+		}
+
+		// scale to fit screen
+		float scale = std::min(w / (float)width, h / (float)height);
```
Comment (john0312) [inline src/DisplayGLWidget.cpp]: Please add const, if possible.

File: src/DisplayGLWidget.cpp [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+		glBegin(GL_QUADS); {
+			glTexCoord2f(0, 1); glVertex2d(-1, -1);
+			glTexCoord2f(0, 0); glVertex2d(-1, 1);
+			glTexCoord2f(1, 0); glVertex2d(1, 1);
+			glTexCoord2f(1, 1); glVertex2d(1, -1);
+
+		} glEnd();
+	}
+
+	// bind vao
```
Comment (john0312) [inline src/DisplayGLWidget.cpp:115]: "TODO"

File: src/DisplayGLWidget.cpp [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+	// draw elements
+
+	// done. unbind
+
+	glDisable(GL_TEXTURE_2D);
+}
+
+void DisplayGLWidget::resizeGL(int width, int height)
+{
+	//[will handle resized event]
```
Comment (john0312) [inline src/DisplayGLWidget.cpp]: Consider calling it "TODO"

File: src/DiplayGLWidget.h [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#ifdef DISPLAY_GLWIDGET_H_
+#define DISPLAY_GLWIDGET_H_
+
+#include <QOpenGLWidget>
+#include <Qtimer>
+#include <opencv2\core.hpp>
```
Comment (john0312) [inline src/DiplayGLWidget.h]: Please use "/"

File: src/DiplayGLWidget.h [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+
+	// Callback function when GLWidget is updated.
+	void paintGL();
+
+	// Callback function when GLWidget is resized.
+	// @param int width
+	// New width.
+	// @param int height
+	// New height.
+	void resizeGL(int width, int height);
```
Comment (john0312) [inline src/DiplayGLWidget.h]: const ref.

File: src/DisplayGLWidget.cpp [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+// SOFTWARE.
+
+#include "DisplayGLWidget.h"
+#include <iostream>
+#include <opencv2/imgproc.hpp>
+
+DisplayGLWidget::DisplayGLWidget(QWidget *parent)
+	: QOpenGLWidget(parent)
+{
+
```
Comment (jeremy5189) [inline src/DisplayGLWidget.cpp]: Add a comment to explain why this is a empty function

File: src/DisplayGLWidget.cpp (hunk 11 above)
Comment (jeremy5189) [reply src/DisplayGLWidget.cpp]: It's not resolved yet

**Review by jeremy5189 — CHANGES_REQUESTED** [review]
As commented

File: src/DisplayGLWidget.h [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#ifndef DISPLAY_GLWIDGET_H_
+#define DISPLAY_GLWIDGET_H_
+
+#include <QOpenGLWidget>
```
Comment (john0312) [inline src/DisplayGLWidget.h:29]: Please order the include.

File: src/DisplayGLWidget.h [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+	// Callback function when GLWidget is resized.
+	// @param int width
+	// New width.
+	// @param int height
+	// New height.
+	void resizeGL(int width, int height);
+	
+private:
+	// Last displayed frame.
+	cv::Mat lastFrame;
```
Comment (john0312) [inline src/DisplayGLWidget.h]: Please fix naming.

**Review by john0312 — APPROVED** [review]

**Comment by john0312** [comment]
Please clang-format before merge.
